substrate/primitives/runtime/src/generic/checked_extrinsic.rs

```diff
@@ -18,81 +18,115 @@
+//   unsigned but includes extension data.
+// TODO: Move usage of `signed` to `format`:
+// - Inherent instead of None.
+// - Signed(id, extension) instead of Some((id, extra)).
+// - Introduce General(extension) for one without a signature.
+
 /// Definition of something that the external world might want to say; its existence implies that it
 /// has been checked and is good, particularly with regards to the signature.
 ///
 /// This is typically passed into [`traits::Applyable::apply`], which should execute
 /// [`CheckedExtrinsic::function`], alongside all other bits and bobs.
 #[derive(PartialEq, Eq, Clone, sp_core::RuntimeDebug)]
-pub struct CheckedExtrinsic<AccountId, Call, Extra> {
+pub struct CheckedExtrinsic<AccountId, Call, Extension> {
 	/// Who this purports to be from and the number of extrinsics have come before
 	/// from the same signer, if anyone (note this is not a signature).
-	pub signed: Option<(AccountId, Extra)>,
+	pub format: ExtrinsicFormat<AccountId, Extension>,
 
 	/// The function that should be called.
 	pub function: Call,
 }
 
-impl<AccountId, Call, Extra, RuntimeOrigin> traits::Applyable
-	for CheckedExtrinsic<AccountId, Call, Extra>
+impl<AccountId, Call, Extension, RuntimeOrigin> traits::Applyable
+	for CheckedExtrinsic<AccountId, Call, Extension>
 where
 	AccountId: Member + MaybeDisplay,
-	Call: Member + Dispatchable<RuntimeOrigin = RuntimeOrigin>,
-	Extra: SignedExtension<AccountId = AccountId, Call = Call>,
+	Call: Member + Dispatchable<RuntimeOrigin = RuntimeOrigin> + Encode,
+	Extension: TransactionExtension<Call, ()>,
 	RuntimeOrigin: From<Option<AccountId>>,
 {
 	type Call = Call;
 
-	fn validate<U: ValidateUnsigned<Call = Self::Call>>(
+	fn validate<I: ValidateUnsigned<Call = Self::Call>>(
 		&self,
-		// TODO [#5006;ToDr] should source be passed to `SignedExtension`s?
-		// Perhaps a change for 2.0 to avoid breaking too much APIs?
 		source: TransactionSource,
 		info: &DispatchInfoOf<Self::Call>,
 		len: usize,
 	) -> TransactionValidity {
-		if let Some((ref id, ref extra)) = self.signed {
-			Extra::validate(extra, id, &self.function, info, len)
-		} else {
-			let valid = Extra::validate_unsigned(&self.function, info, len)?;
-			let unsigned_validation = U::validate_unsigned(source, &self.function)?;
-			Ok(valid.combine_with(unsigned_validation))
+		match self.format {
+			ExtrinsicFormat::Bare => {
+				let inherent_validation = I::validate_unsigned(source, &self.function)?;
+				#[allow(deprecated)]
+				let legacy_validation = Extension::validate_bare_compat(&self.function, info, len)?;
+				Ok(legacy_validation.combine_with(inherent_validation))
+			},
+			ExtrinsicFormat::Signed(ref signer, ref extension) => {
+				let origin = Some(signer.clone()).into();
+				extension.validate_only(origin, &self.function, info, len).map(|x| x.0)
+			},
+			ExtrinsicFormat::General(ref extension) =>
+				extension.validate_only(None.into(), &self.function, info, len).map(|x| x.0),
```

Comment:
That the unsigned validation is not directly lifted to the `TransactionExtension` level is not good. IMO it is now just more confusing than before.

Another issue being that you call here just `validate_only` for the General transaction. How will you find out that a transaction is allowed being a `General` transaction? The old `ValidateUnsigned` logic required explicit approval for calls to be accepted as unsigned transaction. This will now be skipped here.

Reply:
This basically means that I could now DOS the network with `General` transaction that are invalid. (Currently it would be prevented by the broken `transaction-payment` transaction extension. However, this is nothing we should count on.)

Reply:
> That the unsigned validation is not directly lifted to the TransactionExtension level is not good. IMO it is now just more confusing than before.

This is step 2 of [Extrinsic Horizon](https://github.com/paritytech/polkadot-sdk/issues/2415). I think it would have been really difficult to fit these changes in this PR too.

> This basically means that I could now DOS the network with General transaction that are invalid. (Currently it would be prevented by the broken transaction-payment transaction extension.

Well, not *now* because this PR only introduces the framework for `General` transactions, but they are not really supported anywhere else. It's true that the transaction payment extension would need to change to accomodate for `General` transactions as you can't really pay for a transaction without an account. These changes should follow naturally when we do step 2 mentioned above, where we deal with unsigned transactions as well.

Reply:
> but they are not really supported anywhere else

I can just construct them and send them to the chain. Any chain that will not use the tx-payment signed extension will (if they don't have any other schemes against these General transactions) be able to be dosed. Aka I can just send million of these transactions to the nodes and they will happily accept them.